middlewares/logger.go

```diff
@@ -140,9 +173,38 @@ func (fblh frontendBackendLoggingHandler) ServeHTTP(rw http.ResponseWriter, req
 
 	elapsed := time.Now().UTC().Sub(startTime.UTC())
 	elapsedMillis := elapsed.Nanoseconds() / 1000000
-	fmt.Fprintf(fblh.writer, `%s - %s [%s] "%s %s %s" %d %d "%s" "%s" %s "%s" "%s" %dms%s`,
-		host, username, ts, method, uri, proto, status, size, referer, agent, fblh.reqid, frontend, backend, elapsedMillis, "\n")
 
+	if fblh.format == "json" {
+		e := logEntryPool.Get().(*logEntry)
+		defer logEntryPool.Put(e)
+		e.RemoteAddr = host
+		e.Username = username
+		e.Timestamp = ts
+		e.Method = method
+		e.URI = uri
+		e.Protocol = proto
+		e.Status = status
+		e.Size = size
+		e.Referer = referer
+		e.UserAgent = agent
+		e.RequestID = fblh.reqid
+		e.Frontend = frontend
+		e.Backend = backend
+		e.ElapsedMillis = elapsedMillis
+		e.Host = req.Host
+
+		data, err := json.Marshal(e)
+		if err != nil {
+			log.Error("unable to marshal json for log entry", err)
+			return
+		}
+		data = append(data, newLineByte)
+		// must do the write as one action to avoid interleaving newlines, etc
+		fblh.writer.Write(data)
```

Comment:
Everything should be on one line anyway, or the logs are going to be unusable by some(or all) log shipping/etc services.